alibi_detect/od/pytorch/base.py

```diff
@@ -0,0 +1,235 @@
+        x
+            Data to convert.
+        """
+        return torch.as_tensor(x, dtype=torch.float32, device=self.device)
+
+    def _ensembler(self, x: torch.Tensor) -> torch.Tensor:
+        """Aggregates and normalizes the data
+
+        If the detector has an ensembler attribute we use it to aggregate and normalize the data.
+
+        Parameters
+        ----------
+        x
+            Data to aggregate and normalize.
+
+        Returns
+        -------
+        `torch.Tensor` or original data without alteration
+
+        Raises
+        ------
+        ThresholdNotInferredError
+            If the detector is an ensemble, and the ensembler used to aggregate the outlier scores has a fittable
+            component, then the detector threshold must be inferred before predictions can be made. This is because
+            while the scoring functionality of the detector is fit within the `.fit` method on the training data
+            the ensembler has to be fit on the validation data along with the threshold and this is done in the
+            `.infer_threshold` method.
+        """
+        if hasattr(self, 'ensembler') and self.ensembler is not None:
+            # `type: ignore` here because self.ensembler here causes an error with mypy when using torch.jit.script.
+            # For some reason it thinks self.ensembler is a torch.Tensor and therefore is not callable.
+            if not torch.jit.is_scripting():
+                if not self.ensembler.fitted:  # type: ignore
+                    self.check_threshold_inferred()
+            return self.ensembler(x)  # type: ignore
+        else:
+            return x
+
+    def _classify_outlier(self, scores: torch.Tensor) -> torch.Tensor:
+        """Classify the data as outlier or not.
+
+        Parameters
+        ----------
+        scores
+            Scores to classify. Larger scores indicate more likely outliers.
+
+        Returns
+        -------
+        `torch.Tensor` or ``None``
+        """
+        return scores > self.threshold if self.threshold_inferred else None
+
+    def _p_vals(self, scores: torch.Tensor) -> torch.Tensor:
+        """Compute p-values for the scores.
+
+        Parameters
+        ----------
+        scores
+            Scores to compute p-values for.
+
+        Returns
+        -------
+        `torch.Tensor` or ``None``
+        """
+        return (1 + (scores[:, None] < self.val_scores).sum(-1))/len(self.val_scores) \
+            if self.threshold_inferred else None
+
+    def infer_threshold(self, x_ref: torch.Tensor, fpr: float):
+        """Infer the threshold for the data. Prerequisite for outlier predictions.
+
+        Parameters
+        ----------
+        x_ref
+            Data to infer the threshold for.
+        fpr
+            False positive rate to use for threshold inference.
+
+        Raises
+        ------
+        ValueError
+            Raised if `fpr` is not in ``(0, 1)``.
+        """
+        if not 0 < fpr < 1:
```

Comment:
Would be good to have another check here that the fpr is achievable given the number of samples provided. For example the lowest achievable FPR given 100 samples is 0.01.